service/matching/db.go

```diff
@@ -90,6 +92,20 @@ func (db *taskQueueDB) RangeID() int64 {
 	return db.rangeID
 }
 
+// GetMaxReadLevel returns the current maxReadLevel
+func (db *taskQueueDB) GetMaxReadLevel() int64 {
+	db.RLock()
+	defer db.RUnlock()
+	return db.maxReadLevel
```

Comment:
Not a Go expert by any means, but if we do use the same locks (which we are in this case since we are using a RW mutex) for this resource, synchronization of values are also taken care of right?  thought this was a given with using the same locking object but yeah curious to see what @dnr states

Reply:
Yes, this is fine. I have a little concern that now taskReader will get blocked calling GetMaxReadLevel while we're in the db calling db.store.CreateTasks. In practice it may not be a big deal, but it is a regression.

We could avoid it by keeping maxReadLevel as an atomic, having GetMaxReadLevel only load the atomic without a lock, and SetMaxReadLevel get the lock and then write the atomic.

Effectively that's the same as having two mutexes, one just for maxReadLevel and one for everything else, and always taking the "everything else" before the "maxReadLevel"

Reply:
Interesting idea @dnr  - seem to be in agreement since it could speed up things with an atomic. Carrying on with the implementation.

Reply:
on second thought, could you elaborate a little more on this?

Effectively that's the same as having two mutexes, one just for maxReadLevel and one for everything else, and always taking the "everything else" before the "maxReadLevel"

If we do go ahead with this, if `taskReader` calls `GetMaxReadLevel` while `db` is locked and before updating `maxReadLevel`, there is a chance that it could return an older value to it right? Are we okay with this and bank on the fact that `taskReader` will again call `getTaskBatch` and then read the resource again, where it shall have the latest value ?

Reply:
If another goroutine has the write lock, GetMaxReadLevel doesn't return an old value, it just blocks. If it were just an in-memory operation that's might be fine, but in this case the write lock is held across a call to the db.

We do want to hold that lock across the db call as a way to serialize db operations (necessary on cassandra at least). So we can't really shrink that lock. But what we could do is put just maxReadLevel under a different lock so we can read it without that large db lock. That's basically how it was before.. you can think of an atomic as just a single value plus a lock just for that value.

Reply:
Yeah I think we should be fine doing what @dnr suggests. I also did worry a little about locking read path. but did not bring it up thinking we read batches so reads are infrequent and wouldn't be highly impacted. But I don't think that's a valid assumption in may cases where a small backlog is formed for a little while because the poller came a few milliseconds late, and maybe one or two tasks are read from DB every time.

Nevertheless, I think what David is suggesting is good because we don't have to block GetMaxReadLevel to solve the [original problem](https://github.com/temporalio/temporal/pull/5593/files/ecbbac9dd6e62433aaec4069438d65e5c78148ec#r1552145429) which is only about synchronizing UpdateState and CreateTasks.